python/packages/autogen-ext/src/autogen_ext/models/anthropic/_anthropic_client.py

```diff
@@ -0,0 +1,987 @@
+    def _from_config(cls, config: AnthropicClientConfigurationConfigModel) -> Self:
+        copied_config = config.model_copy().model_dump(exclude_none=True)
+        return cls(**copied_config)
+    Normalize names by replacing invalid characters with underscore.
+    """
+    return re.sub(r"[^a-zA-Z0-9_-]", "_", name)[:64]
+
+
+def assert_valid_name(name: str) -> str:
+    """
+    Ensure that configured names are valid, raises ValueError if not.
+    """
+    if not re.match(r"^[a-zA-Z0-9_-]+$", name):
+        raise ValueError(f"Invalid name: {name}. Only letters, numbers, '_' and '-' are allowed.")
+    if len(name) > 64:
+        raise ValueError(f"Invalid name: {name}. Name must be less than 64 characters.")
+    return name
+
+
+def normalize_stop_reason(stop_reason: str | None) -> FinishReasons:
+    if stop_reason is None:
+        return "unknown"
+
+    # Convert to lowercase for comparison
+    stop_reason = stop_reason.lower()
+
+    # Map Anthropic stop reasons to standard reasons
+    KNOWN_STOP_MAPPINGS: Dict[str, FinishReasons] = {
+        "end_turn": "stop",
+        "max_tokens": "length",
+        "stop_sequence": "stop",
+        "tool_use": "function_calls",
+    }
+
+    return KNOWN_STOP_MAPPINGS.get(stop_reason, "unknown")
+
+
+def _add_usage(usage1: RequestUsage, usage2: RequestUsage) -> RequestUsage:
+    return RequestUsage(
+        prompt_tokens=usage1.prompt_tokens + usage2.prompt_tokens,
+        completion_tokens=usage1.completion_tokens + usage2.completion_tokens,
+    )
+
+
+class BaseAnthropicChatCompletionClient(ChatCompletionClient):
+    def __init__(
+        self,
+        client: AsyncAnthropic,
+        *,
+        create_args: Dict[str, Any],
+        model_capabilities: Optional[ModelCapabilities] = None,  # type: ignore
+        model_info: Optional[ModelInfo] = None,
+    ):
+        self._client = client
+
+        # Handle model info and capabilities
+        if model_capabilities is None and model_info is None:
+            try:
+                self._model_info = _model_info.get_info(create_args["model"])
+            except KeyError as err:
+                raise ValueError("model_info is required when model name is not recognized") from err
+        elif model_capabilities is not None and model_info is not None:
+            raise ValueError("model_capabilities and model_info are mutually exclusive")
+        elif model_capabilities is not None and model_info is None:
+            warnings.warn("model_capabilities is deprecated, use model_info instead", DeprecationWarning, stacklevel=2)
+            info = cast(ModelInfo, model_capabilities)
+            info["family"] = ModelFamily.CLAUDE_3_5_SONNET
+            self._model_info = info
+        elif model_capabilities is None and model_info is not None:
+            self._model_info = model_info
+
+        # Validate model_info
+        if not all(key in self._model_info for key in ["vision", "function_calling", "json_output", "family"]):
```

Comment:
There is a helper function autogen_core.models.validate_model_info that does the same thing.

Reply:
Added